pmda/rmsf.py

```diff
@@ -0,0 +1,216 @@
+    whole trajectory to the reference by minimizing the RMSD. We use
+    :class:`MDAnalysis.analysis.align.AlignTraj`::
+
+       aligner = align.AlignTraj(u, reference, select="protein and name CA",
+                                 in_memory=True).run()
+
+    The trajectory is now fitted to the reference (the RMSD is stored as
+    `aligner.rmsd` for further inspection). Now we can calculate the RMSF::
+
+       from MDAnalysis.analysis.rms import RMSF
+       calphas = protein.select_atoms("name CA")
+       rmsfer = RMSF(calphas, verbose=True).run()
+
+    and plot::
+
+       import matplotlib.pyplot as plt
+       plt.plot(calphas.resnums, rmsfer.rmsf)
+
+    References
+    ----------
+    .. [Welford1962] B. P. Welford (1962). "Note on a Method for
+       Calculating Corrected Sums of Squares and Products." Technometrics
+       4(3):419-420.
+
+    See Also
+    --------
+    MDAnalysis.analysis.density.density_from_Universe
+
+
+    .. versionadded:: 0.3.0
+
+    """
+    def __init__(self, atomgroup):
+        u = atomgroup.universe
+        super(RMSF, self).__init__(u, (atomgroup, ))
+        self._atomgroup = atomgroup
+        self._top = u.filename
+        self._traj = u.trajectory.filename
+
+    def _prepare(self):
+        self.sumsquares = np.zeros((self._atomgroup.n_atoms, 3))
+        self.mean = self.sumsquares.copy()
+        # print("Sum of squares array: ", self.sumsquares)
+        # print("Mean array: ", self.mean)
+
+    def _single_frame(self, ts, agroups):
+        k = self._trajectory.frame
+        self.sumsquares += (k / (k + 1)) * (agroups[0].positions - self.mean) ** 2
+        self.mean = (k * self.mean + agroups[0].positions) / (k + 1)
+
+    def _conclude(self):
+        """
+        self._results : Array
+            (n_blocks x ts x 2 x N x 3) array
+            (1 x 901 x 2 x 10 x 3)
+        """
+        k = len(self._trajectory)
+        self.sumsquares = self._results[0, :, 0]
+        self.mean = self._results[0, :, 1]
+        self.rmsf = np.sqrt(self.sumsquares.sum(axis=0) / k)
+        if not (self.rmsf >= 0).all():
+            raise ValueError("Some RMSF values negative; overflow " +
+                             "or underflow occurred")
+
+    @staticmethod
+    def _reduce(res, result_single_frame):
+        """
+        'append' action for a time series
+        """
+        res.append(result_single_frame)
+        return res
+
+    # def _reduce(res, result_single_frame):
+    #     res.append(results_single_frame)
+    #     if res == []:
+    #         res = result_single_frame
+    #     else:
+    #         res += result_single_frame
+    #     return res
+
+    @staticmethod
+    def pair_wise_rmsf(mu1, mu2, t1, t2, M1, M2):
+        """
+        Calculates the total RMSF pair-wise. Takes in two separate blocks with
+        after the RMSF calculation has been concluded and combines their results
+        into a single, total RMSF for the combined trajectory slices.
+
```

Comment:
Need to add reference for algorithm.

See https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#guidelines and in particular https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard for the order of sections (implementation/algorithm is in Notes and the end and References is the very last section)